### PR TITLE
Paris revision support

### LINKED ISSUE #388: evmzero: support Paris revision
Support for the instructions added by the Paris revision to the evmzero implementation needs to be added.

### PR DESCRIPTION
**Pull request by simonlechner**
This PR adds support for the Paris revision to the CT, LFVM and evmzero.
The only change to the VM for this revision has been the renaming from DIFFICULTY to PREVRANDAO. 
Fixes #388, #391, #414 and #464.

### CODE REVIEW

**Review by facuMH — DISMISSED** [review]
LGTM

File: cpp/vm/evmzero/interpreter_test.cc [hunk 1]
```diff
@@ -2978,15 +2978,15 @@ TEST(InterpreterTest, DIFFICULTY) {
...
   RunInterpreterTest({
-      .code = {op::DIFFICULTY},
+      .code = {op::PREVRANDAO},
       .state_after = RunState::kErrorGas,
       .gas_before = 1,
   });
 }
 
-TEST(InterpreterTest, DISABLED_DIFFICULTY_StackOverflow) {}
+TEST(InterpreterTest, DISABLED_PREVRANDAO_StackOverflow) {}
```
Comment (HerbertJordan) [inline cpp/vm/evmzero/interpreter_test.cc:2989]: This test is useless -- either we fix it or remove it.

File: go/ct/driver/run.go [hunk 2]
```diff
@@ -228,7 +228,9 @@ func runTest(input *st.State, evm ct.Evm, filter *regexp.Regexp) error {
 	rule.Effect.Apply(expected)
 
 	result, err := evm.StepN(input.Clone(), 1)
-	defer result.Release()
+	if err == nil {
+		defer result.Release()
+	}
```
Comment (HerbertJordan) [inline go/ct/driver/run.go:233]: nit: Better to move the `defer` statement after the error check below.

**Review by HerbertJordan — APPROVED** [review]

File: cpp/vm/evmzero/interpreter_test.cc (hunk 1 above)
Comment (simonlechner) [reply cpp/vm/evmzero/interpreter_test.cc:2989]: There are 14 other disabled stack overflow tests, I think they should be handled in a separate PR.

**Review by HerbertJordan — APPROVED** [review]

File: go/ct/common/opcodes.go [hunk 3]
```diff
@@ -292,8 +292,8 @@ func (op OpCode) String() string {
 		return "TIMESTAMP"
 	case NUMBER:
 		return "NUMBER"
-	case DIFFICULTY:
-		return "DIFFICULTY"
+	case PREVRANDAO:
+		return "PREVRANDAO"
```
Comment (LuisPH3) [inline go/ct/common/opcodes.go:296]: The code 44 still translates to DIFFICULTY for revisions < Paris,
I guess we should respect the printouts.

File: go/ct/driver/run.go [hunk 4]
```diff
@@ -228,10 +228,10 @@ func runTest(input *st.State, evm ct.Evm, filter *regexp.Regexp) error {
 	rule.Effect.Apply(expected)
 
 	result, err := evm.StepN(input.Clone(), 1)
-	defer result.Release()
 	if err != nil {
 		return err
 	}
+	defer result.Release()
```
Comment (LuisPH3) [inline go/ct/driver/run.go:234]: what this a bug before?  why does this change now?

File: go/vm/lfvm/converter.go [hunk 5]
```diff
@@ -528,7 +528,7 @@ func createOpToOpMap() []OpCode {
 	res[evm.COINBASE] = COINBASE
 	res[evm.TIMESTAMP] = TIMESTAMP
 	res[evm.NUMBER] = NUMBER
-	res[evm.DIFFICULTY] = DIFFICULTY
+	res[evm.DIFFICULTY] = PREVRANDAO //TODO: update with new geth version
```
Comment (LuisPH3) [inline go/vm/lfvm/converter.go:531]: is todo still valid?

**Review by LuisPH3 — APPROVED** [review]
I guess that the to string function should print the old name of the instructions when reading a program with revision < paris. 

Nevertheless I understand that the revision is not available at this point and having this cosmetic feature could be rather difficult to implement.

**Review by facuMH — APPROVED** [review]

File: go/ct/common/opcodes.go (hunk 3 above)
Comment (simonlechner) [reply go/ct/common/opcodes.go:296]: As discussed the opcode name is not used in any important functionality therefore it is sufficient to just return the new name.

File: go/ct/driver/run.go (hunk 4 above)
Comment (simonlechner) [reply go/ct/driver/run.go:234]: Yes, this should have always been performed after the error check.

File: go/vm/lfvm/converter.go (hunk 5 above)
Comment (simonlechner) [reply go/vm/lfvm/converter.go:531]: Yes, once the geth version is updated this line changes to `res[evm.PREVRANDAO] = PREVRANDAO`.

**Review by facuMH — APPROVED** [review]

**Pull request merged**